File: reboot/examples/bank-zod/backend/src/customer.ts
import {
  allow,
  ReaderContext,
  TransactionContext,
  WriterContext,
} from "@reboot-dev/reboot";
import { v4 as uuidv4 } from "uuid";
import { Account } from "../../api/bank/v1/account_rbt.js";
import { Customer } from "../../api/bank/v1/customer_rbt.js";

export const CustomerServicer = Customer.servicer({
  authorizer: () => {
    return allow();
  },

  signUp: async (
    context: WriterContext,
    state: Customer.State,
    request: Customer.SignUpRequest
  ): Promise<[Customer.State, Customer.PartialSignUpResponse]> => {
    return [state, {}];
  },

  openAccount: async (
    context: TransactionContext,
    state: Customer.State,
    { initialDeposit }: Customer.OpenAccountRequest
  ): Promise<[Customer.State, Customer.PartialOpenAccountResponse]> => {
    const accountId = uuidv4();
    state.accountIds.push(accountId);

    const [account] = await Account.open(context, accountId, {});

    await account.deposit(context, { amount: initialDeposit });

    return [state, { accountId }];
  },

  balances: async (
    context: ReaderContext,
    state: Customer.State,
    request: Customer.BalancesRequest
  ): Promise<Customer.PartialBalancesResponse> => {
    let balances = [];
    for (const accountId of state.accountIds) {
      const balance = await Account.ref(accountId).balance(context);
      balances.push({ accountId, balance: balance.amount });
    }
    return { balances };
  },
});

File: reboot/examples/bank-pydantic/api/bank/v1/customer.py
from reboot.api import (
    API,
    Exclusive,
    Field,
    Methods,
    Model,
    Reader,
    Transaction,
    Type,
    Writer,
)


class CustomerState(Model):
    account_ids: list[str] = Field(
        tag=1,
        default_factory=list,
        description="The ids of the customer's accounts, in the order "
        "they were opened.",
    )


class OpenAccountRequest(Model):
    initial_deposit: float = Field(tag=1)


class OpenAccountResponse(Model):
    account_id: str = Field(tag=1)


class Balance(Model):
    account_id: str = Field(tag=1)
    balance: float = Field(tag=2)


class BalancesResponse(Model):
    balances: list[Balance] = Field(tag=1, default_factory=list)


CustomerMethods = Methods(
    sign_up=Writer(
        request=None,
        response=None,
        factory=True,
        description="Bring the customer into existence with no "
        "accounts.",
        mcp=None,
    ),
    open_account=Transaction(
        mode=Exclusive(),
        request=OpenAccountRequest,
        response=OpenAccountResponse,
        description="Open an account for this customer with an "
        "initial deposit, returning the id it was given.",
        mcp=None,
    ),
    balances=Reader(
        request=None,
        response=BalancesResponse,
        description="The balance of every account this customer owns.",
        mcp=None,
    ),
)

api = API(
    Customer=Type(
        state=CustomerState,
        methods=CustomerMethods,
        description="One customer, and the accounts they own.",
    ),
)
